### DISCUSSION TITLE
ValueError: Using HILO

### DISCUSSION THREAD

**Discussion by wissamkafa** [hilo-mpc/hilo-mpc#26] [Q&A] [answered]
Hi,
First of all, thank you for this wonderful library and the efforts you put into it.
I am interested in using NN to improve traditional controller performance.
I started to use HILO.
I have all packages installed, and following the code written in the documentation:
https://hilo-mpc.github.io/hilo-mpc/examples/nmpc_hybrid_bio.html
I tried to test the first code.
- First in (from hilo_mpc.library.models import  ecoli_D1210_conti_plant) the "ecoli_D1210_conti_plant" is not reocgnizable instead I used "ecoli_D1210_conti" --- are they the same? or there is something missing here?
- Second: the code runs without problems until the line "control_loop.run(100, p=[100, 4])"
here we get an error:
/home/user/anaconda3/envs/HILO/lib/python3.9/site-packages/hilo_mpc/modules/dynamic_model/dynamic_model.py", line 3965, in simulate
    raise ValueError(f"Dimension mismatch. Supplied dimension for the parameters 'p' is "
ValueError: Dimension mismatch. Supplied dimension for the parameters 'p' is 2x1, but required dimension is 5x1.

it looks like a dimensions mismatch, and to test this I passed [100, 4, 0, 0, 0] instead [100, 4]
but another error raises:
File "/home/user/anaconda3/envs/HILO/lib/python3.9/site-packages/hilo_mpc/modules/controller/mpc.py", line 763, in optimize
    raise ValueError(
ValueError: The model has 2 constant parameter(s): ['Sf', 'If']. You must pass me the value of these before running the optimization to the 'cp' parameter.
I also tried to assign values for Sf and If before calling control_loop.run() but without any improvement.

Any idea why this is happening, am I missing something or what?

**Comment by brunomorampc** [comment]
Hi @wissamkafa ,

I apologize for the late reply. Somehow I did not get notifications when someone adds a post to the discussions. 

I created an issue, I'll look into the issue today/tomorrow and let you know.

**Reply by wissamkafa** [reply]
Thanks for your interest. @brunomorampc 
Next time I will post an issue instead of a question here.

Actually, I am interested in using HILO in my research. I am a Ph.D. student working with control systems and neural networks and found your library as a good start.

Currently, I am reading the documentation and the provided examples.
I wish it has more detailed documentation, I find it a bit difficult to understand what's going on. (So, maybe later you will see more questions from me)

For now, can you please confirm if a system represented by nonlinear (second or third) order differential equations could be implemented, simulated, and controlled using MPC and NN?

**Reply by brunomorampc** [reply]
Yes, you can solve any order of differential equations, you need to transform them though into a system of differential equations of order 1 and then it's done.

Yes please let us know what part of the documentation is not clear. We appreciate any feedback and we'll try to fix the issues as soon as possible

**Reply by brunomorampc** [reply, marked as answer]
Dear @wissamkafa,

I fixed the documentation. The problems you reported where do the fact that the model loaded from the library was wrong. 

In general, it is better to look at our example https://github.com/hilo-mpc/examples there you find the same examples as in the documentation but updated to the last version

**Reply by wissamkafa** [reply]
@brunomorampc 
Thanks for your interest, 
Now I am working with the examples
